content/go/concepts/interfaces/interfaces.md

```diff
@@ -0,0 +1,270 @@
+}
+
+func main() {
+    newLife(&john{})
+    newLife(&max{})
+}
+
+```
+
+Output:
+
+```bash
+
+love with john
+arranged marriage with john
+love with max
+arranged marriage with max
+
+```
+
+In the above example,
+
+- The `marriage` interface has two methods `love` and `arranged`.
+- Structs `john` and `max` implement the `marriage` interface. Note that neither `john` nor `max` has any explicit declaration to implement the interface.
+- As long as the struct has all the methods in the interface, it is said to implement the interface.
+- The function `newLife` takes a `marriage` interface as an argument.
+- It's possible to call the `newLife` function with the `john` and `max` structs.
+
+The advantage of using an interface is that any struct can be passed to the `newLife` function as long as it implements the `marriage` interface. It is possible to change the implementation of the `love` and `arranged` methods in the `john` and `max` structs without affecting the `newLife` function.
+
+It is also possible to make a new type and `newLife` can use it without changing the `newLife` function as long as the new type implements the `marriage` interface.
+
+Go's compiler will ensure that all the methods in the interface are implemented in the struct.
+
+## Empty Interface
+
+An interface with no methods is called an empty interface.
+
+```go
+type interfaceName interface {}
+```
+
+The tradeoff of using an empty interface is the loss of type safety. As there are no methods, it's possible to pass any type to a function that takes an empty interface as an argument.
+
+### Example
+
+The `Println` function in the `fmt` package takes an empty interface as an argument.
+
+Below is the go source code for the `fmt.Println` function:
+
+```go
+// Println formats using the default formats for its operands and writes to standard output.
+// Spaces are always added between operands and a newline is appended.
+// It returns the number of bytes written and any write error encountered.
+func Println(a ...interface{}) (n int, err error) {
+    return Fprintln(os.Stdout, a...)
+}
+```
+
+> ##Note:\*\* due to the empty interface, we can pass any type to the `Println` function.
+
+## Embedding an Interface
+
+An interface in go can be embedded in other interfaces or structs.
+
+```go
+package main
+
+import "fmt"
+
+// any type that need to implement marriage should also implement all methods of dating
+type marriage interface {
+    dating
+    love()
+    arranged()
+}
+
+type dating interface {
+    compatibility()
+}
+
+type john struct {
+    age int
+}
+
+func (a john) love() {
+    fmt.Println("love with john")
+}
+
+func (a john) arranged() {
+    fmt.Println("arranged marriage with john")
+}
+func (a john) compatibility() {
+    fmt.Println("compatible with john")
+}
+
+type parents struct {
+    //named marriage interface
+    m     marriage
+    quote string
+}
+
+type lovers struct {
+    //unnamed marriage interface
+    marriage
+    quote string
```

Comment:
Fix any typos here and capitalize comments

Reply:
ok